plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,538 @@
+
+	fw_bytes = g_bytes_get_data(fw, NULL);
+	fw_total_size = g_bytes_get_size(fw);
+	fw_size = fw_total_size / 2;
+
+	// TODO: Do this in prepare_firmware
+	// Make sure there are two firmware images composed of full rows
+	g_assert(fw_total_size % 2 * ROW_SIZE == 0);
+	fw1_binary = fw_bytes;
+	fw2_binary = &fw_bytes[fw_size];
+
+	fu_ccgx_pure_hid_magic_unlock(device, NULL, error);
+	if (!fu_ccgx_pure_hid_get_fw_info_wrapper(device, NULL, error)) {
+		g_prefix_error(error, "get fw info error: ");
+		return FALSE;
+	}
+
+	g_debug("Operating Mode:  0x%02x", self->operating_mode);
+	switch (self->operating_mode) {
+		case FU_CCGX_FW_MODE_BOOT:
+		case FU_CCGX_FW_MODE_FW2:
+			// Update Image 1
+			g_debug("Flashing Image 1");
+			fu_ccgx_pure_hid_flash_firmware_image(device, fw1_binary, fw_size, FW1_START, FW1_METADATA, 1, error);
+
+			g_debug("Add wait for replug");
+			fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+			fu_device_add_private_flag(device, FU_CCGX_PURE_HID_DEVICE_IS_IN_RESTART);
+
+			// TODO: Wait 5s
+			// TODO: Refresh HID devices
+			//fu_ccgx_pure_hid_magic_unlock(device, user_data, error);
+			// Update Image 2
+			// flash_firmware_image(device, fw2_binary, fw_size, FW2_START, FW2_METADATA, 2);
+			break;
+		case FU_CCGX_FW_MODE_FW1:
+			// Update Image 2
+			g_debug("Flashing Image 2");
+			fu_ccgx_pure_hid_flash_firmware_image(device, fw2_binary, fw_size, FW2_START, FW2_METADATA, 2, error);
+
+			g_debug("Add wait for replug");
+			fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+			fu_device_add_private_flag(device, FU_CCGX_PURE_HID_DEVICE_IS_IN_RESTART);
+
+			// TODO: Wait 5s
+			// TODO: Refresh HID devices
+			//fu_ccgx_pure_hid_magic_unlock(device, user_data, error);
+			// Update Image 1
+			//flash_firmware_image(device, fw1_binary, fw_size, FW1_START, FW1_METADATA, 1);
+			break;
+		default:
+			// Invalid value
+			g_assert(self->operating_mode < FU_CCGX_FW_MODE_LAST);
+			break;
+	}
+
+	return TRUE;
+}
+
+static void
+fu_ccgx_pure_hid_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 25, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 50, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 25, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 50, "reload");
+}
+
+static void
+fu_ccgx_pure_hid_device_init(FuCcgxPureHidDevice *self)
+{
+	fu_device_add_protocol(FU_DEVICE(self), "com.cypress.ccgx");
+	fu_device_add_protocol(FU_DEVICE(self), "com.infineon.ccgx");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UNSIGNED_PAYLOAD);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_INTEL_ME2);
```

Comment:
is this the right version format?

Reply:
> I've made a script to generate the cyacd file like fwupd expects it:

I don't think you should modify the firmware just for fwupd; I think the plugin needs to adapt - in this case just using the BaseVersion for display, not the AppVersion. e.g. have a private flag that is called something like `prefer-base-version` or something.

I can play with this code tomorrow; sorry for nor playing with your branch already -- I'm juggling vendors like crazy.

Reply:
You can do; but lets see what we can do by modifying fu-ccgx-firmware.c: What offset is base-version? app-version is 228 if that helps.

Reply:
Also, can you upload the HDMI-6-1.cyacd file somewhere (or send me an email please)